x/stakeibc/keeper/unbonding_records.go

```diff
@@ -96,6 +111,74 @@ func (k Keeper) GetTotalUnbondAmountAndRecordsIds(ctx sdk.Context, chainId strin
 	return totalUnbonded, unbondingRecordIds
 }
 
+// Update Native Amounts in the Host Zone Unbonding Record (and associated User Redemption Records) to reflect the current redemption rate
+// If setNativeAmountsToZero is true, then the native amounts are set to zero, ignoring the redemption rate. This should be used if the unbonding didn't succeed
+// Otherwise, the native amounts are set based on the current redemption rate
+func (k Keeper) UpdateNativeTokensForHostZoneUnbondingRecord(
+	ctx sdk.Context,
+	epochNumber uint64,
+	hostZoneUnbondingRecord *recordstypes.HostZoneUnbonding,
+	setNativeAmountsToZero bool,
+) bool {
+	totalNativeAmount := sdkmath.ZeroInt()
+	chainId := hostZoneUnbondingRecord.GetHostZoneId()
+	hostZone, found := k.GetHostZone(ctx, chainId)
+	if !found {
+		k.Logger(ctx).Info(utils.LogWithHostZone(chainId, "No host zone found for id %s, Not Updating Native Token Amounts", chainId))
+		return false
+	}
+	redemptionRate := hostZone.RedemptionRate
+	// loop through user redemption records for the host zone unbonding record, update each one
+	for _, userRedemptionRecordId := range hostZoneUnbondingRecord.GetUserRedemptionRecords() {
+		userRedemptionRecord, found := k.RecordsKeeper.GetUserRedemptionRecord(ctx, userRedemptionRecordId)
+		if !found {
+			k.Logger(ctx).Info(utils.LogWithHostZone(chainId, "No user redemption record found for id %s", userRedemptionRecordId))
+			continue
+		}
+		// update the amount, based on the current redemption rate (or set to zero if desired)
+		nativeAmount := sdkmath.ZeroInt()
+		if !setNativeAmountsToZero {
+			nativeAmount = sdk.NewDecFromInt(userRedemptionRecord.StTokenAmount).Mul(redemptionRate).RoundInt()
+		}
+		userRedemptionRecord.Amount = nativeAmount
+		totalNativeAmount = totalNativeAmount.Add(nativeAmount)
+		k.RecordsKeeper.SetUserRedemptionRecord(ctx, userRedemptionRecord)
+	}
+	// update the host zone unbonding record
+	hostZoneUnbondingRecord.NativeTokenAmount = totalNativeAmount
+	updatedEpochUnbondingRecord, success := k.RecordsKeeper.AddHostZoneToEpochUnbondingRecord(ctx, epochNumber, chainId, hostZoneUnbondingRecord)
+	if !success {
+		// TODO QUESTION: what should we do on failure? do we revert the userRedemptionRecord changes above?
```

Comment:
Yeah that makes sense! The reason I threw an error in the first location is because we need that function call to succeed (otherwise we don't know how many tokens to unbond), but we don't strictly need it in the other function call (it's only used for housekeeping niceness). 

That being said, I lean towards throwing an error regardless. I put the tradeoffs in [here](https://github.com/Stride-Labs/stride/pull/1049#discussion_r1446777459), my recommendation is we get one-notch more confident that the function won't fail, and then we choose to always revert.

Reply:
Added a comments below at the [link](https://github.com/Stride-Labs/stride/pull/1049#discussion_r1446777459). (Low confidence) I think the records could get out of sync if we don't revert. Agree we should get one-notch more confident that the function won't fail, and then we choose to always revert.

Reply:
Even if we throw an error, state doesn't get reverted since this is in the begin blocker. We should use a cache context wrapper around this function and throw an error

Reply:
it's not necessarily in the begin blocker - re-opening ICAs is just a tx